drivers/sensor/scd30/scd30.c

```diff
@@ -0,0 +1,267 @@
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	float co2_ppm, temp, rel_hum;
+
+	co2_ppm = scd30_bytes_to_float(data_buf);
+	temp = scd30_bytes_to_float(data_buf + 4);
+	rel_hum = scd30_bytes_to_float(data_buf + 8);
+
+	sensor_value_from_double(&data->co2_ppm, co2_ppm);
+	sensor_value_from_double(&data->temp, temp);
+	sensor_value_from_double(&data->rel_hum, rel_hum);
+
+	return 0;
+}
+
+static int scd30_channel_get(const struct device *dev, enum sensor_channel chan,
+			     struct sensor_value *val)
+{
+	struct scd30_data *data = dev->data;
+
+	switch (chan) {
+	case SENSOR_CHAN_CO2:
+		memcpy(val, &data->co2_ppm, sizeof(struct sensor_value));
+		break;
+
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		memcpy(val, &data->temp, sizeof(struct sensor_value));
+		break;
+
+	case SENSOR_CHAN_HUMIDITY:
+		memcpy(val, &data->rel_hum, sizeof(struct sensor_value));
+		break;
+
+	default:
+		return -1;
+	}
+	return 0;
+}
+
+static const struct sensor_driver_api scd30_driver_api = {
+	.sample_fetch = scd30_sample_fetch,
+	.channel_get = scd30_channel_get,
+};
+
+int scd30_write_command(const struct device *dev, uint16_t cmd)
+{
+	uint8_t tx_buf[2] = { cmd >> 8, cmd & 0xFF };
+
+	return i2c_write(scd30_i2c_device(dev), tx_buf, sizeof(tx_buf), scd30_i2c_address(dev));
+}
+
+static uint8_t scd30_compute_crc(uint8_t *data, uint8_t data_len)
+{
+	return crc8(data, data_len, SCD30_CRC8_POLYNOMIAL, SCD30_CRC8_INIT, false);
+}
+
+int scd30_check_crc(uint8_t *data, uint8_t data_len, uint8_t checksum)
+{
+	uint8_t actual_crc = scd30_compute_crc(data, data_len);
+
+	if (checksum != actual_crc) {
+		LOG_DBG("CRC check failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int scd30_write_register(const struct device *dev, uint16_t cmd, uint16_t val)
+{
+	uint8_t tx_buf[SCD30_CMD_SINGLE_WORD_BUF_LEN];
+
+	tx_buf[0] = cmd >> 8;
+	tx_buf[1] = cmd & 0xFF;
+	tx_buf[2] = val >> 8;
+	tx_buf[3] = val & 0xFF;
+	tx_buf[4] = scd30_compute_crc(tx_buf, sizeof(tx_buf));
+
+	return i2c_write(scd30_i2c_device(dev), tx_buf, ARRAY_SIZE(tx_buf), scd30_i2c_address(dev));
+}
+
+static int scd30_init(const struct device *dev)
+{
+	LOG_DBG("Initializing SCD30");
+	struct scd30_data *data = dev->data;
+	const struct scd30_config *cfg = dev->config;
+	const struct device *i2c = device_get_binding(cfg->bus_name);
+
+	if (i2c == NULL) {
+		LOG_DBG("Failed to get pointer to %s device!", log_strdup(cfg->bus_name));
+		return -EINVAL;
+	}
+	data->bus = i2c;
+
+	if (!cfg->base_address) {
+		LOG_DBG("No I2C address");
+		return -EINVAL;
+	}
+	data->dev = dev;
+
+	LOG_DBG("Starting periodic measurements");
+	int rc = scd30_write_command(dev, SCD30_CMD_START_PERIODIC_MEASUREMENT);
+
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	LOG_DBG("Setting measurement interval");
+	rc = scd30_write_register(dev, SCD30_CMD_SET_MEASUREMENT_INTERVAL,
+				  CONFIG_SCD30_MEASUREMENT_INTERVAL);
+
+	return rc;
+}
+
+DEVICE_DT_INST_DEFINE(0, scd30_init, device_pm_control_nop, &scd30_0_data, &scd30_0_cfg,
```

Comment:
Please:

- use i2c_dt_spec for managing the bus device and address (and use device_is_ready properly in the init function to check the bus is ready before using it)
- support multiple instances using DT_INST_FOREACH_STATUS_OKAY